File: src/main/java/com/jennifertellez/library/service/ShelfService.java
package com.jennifertellez.library.service;

import com.jennifertellez.library.dto.CreateShelfRequest;
import com.jennifertellez.library.dto.ShelfResponse;
import com.jennifertellez.library.dto.UpdateShelfRequest;

import java.util.List;

public interface ShelfService {

    ShelfResponse createShelf(CreateShelfRequest request);

    ShelfResponse getShelfById(Long id);

    List<ShelfResponse> getAllShelves();

    ShelfResponse updateShelf(Long id, UpdateShelfRequest request);

    void deleteShelf(Long id);

    ShelfResponse addBookToShelf(Long shelfId, Long bookId);

    ShelfResponse removeBookFromShelf(Long shelfId, Long bookId);
}

File: library-frontend/src/services/shelfService.ts
import { api } from './api';
import { Shelf, CreateShelfRequest, UpdateShelfRequest } from '../types/shelf';

export const shelfService = {
  // Get all shelves
  getAllShelves: async () => {
    const response = await api.get<Shelf[]>('/shelves');
    return response.data;
  },

  // Get shelf by ID
  getShelfById: async (id: number) => {
    const response = await api.get<Shelf>(`/shelves/${id}`);
    return response.data;
  },

  // Create shelf
  createShelf: async (shelf: CreateShelfRequest) => {
    const response = await api.post<Shelf>('/shelves', shelf);
    return response.data;
  },

  // Update shelf
  updateShelf: async (id: number, shelf: UpdateShelfRequest) => {
    const response = await api.put<Shelf>(`/shelves/${id}`, shelf);
    return response.data;
  },

  // Delete shelf
  deleteShelf: async (id: number) => {
    await api.delete(`/shelves/${id}`);
  },

  // Add book to shelf
  addBookToShelf: async (shelfId: number, bookId: number) => {
    const response = await api.post<Shelf>(`/shelves/${shelfId}/books/${bookId}`);
    return response.data;
  },

  // Remove book from shelf
  removeBookFromShelf: async (shelfId: number, bookId: number) => {
    const response = await api.delete<Shelf>(`/shelves/${shelfId}/books/${bookId}`);
    return response.data;
  },
};